client/network/src/discovery.rs

```diff
@@ -206,15 +204,9 @@ impl DiscoveryConfig {
 			discovery_only_if_under_num,
 			#[cfg(not(target_os = "unknown"))]
 			mdns: if enable_mdns {
-				match Mdns::new() {
-					Ok(mdns) => Some(mdns).into(),
-					Err(err) => {
-						warn!(target: "sub-libp2p", "Failed to initialize mDNS: {:?}", err);
-						None.into()
-					}
-				}
+				MdnsWrapper::Instantiating(Mdns::new().boxed())
```

Comment:
Just a thought: If possible, it may be preferable to incorporate this kind of lazy initialisation into the `Mdns` behaviour itself, so that `Mdns::new()` is no longer async.

Reply:
Only having given this some thought, I am in favor of `Mdns::new()` being `async` instead of being synchronous thus having to delay initialization. By being `async` the caller can handle initialization errors at construction time, instead of having to handle them at runtime.

> However in the absolute I think it makes sense for NetworkWorker::new to be async, even if the caller has to do futures::executor::block_on.

As @tomaka said above I would be in favor of making the whole caller chain `async` instead. But I would prefer this to happen in a separate pull request.